x-pack/plugins/security_solution/server/lib/detection_engine/signals/utils.ts

```diff
@@ -661,3 +663,356 @@ export const createTotalHitsFromSearchResult = ({
       : searchResult.hits.total.value;
   return totalHits;
 };
+
+export interface TimestampsAndIndices {
+  [timestampString: string]: string[]; // maps timestampString like @timestamp or 'event.ingested' to the indices that contain that timestamp mapping
+}
+
+export const checkIndexMappingsForTimestampFields = async (
+  indices: string[],
+  timestamps: string[],
+  services: AlertServices,
+  logger: Logger,
+  buildRuleMessage: BuildRuleMessage
+): Promise<TimestampsAndIndices> => {
+  try {
+    const foundMappings: GetFieldMappingType = await services.callCluster(
+      'indices.getFieldMapping',
+      {
+        index: indices,
+        fields: timestamps,
+      }
+    );
+
+    // get the full names of the indices found to contain the given field mapping
+    const matchedIndices = Object.keys(foundMappings);
+
+    // map the timestamp fields like '@timestamp', 'event.ingested' etc.. to the indices that contain mappings for these fields
+    const toReturn = timestamps.reduce((acc, timestamp) => {
+      return {
+        [timestamp]: matchedIndices.filter(
+          (index) => foundMappings[index]?.mappings[timestamp] != null
+        ),
+        ...acc,
+      };
+    }, {} as TimestampsAndIndices);
+    return toReturn;
+  } catch (exc) {
+    logger.error(
+      buildRuleMessage(
+        `An error occurred while getting the field mapping type for this rule ${exc}`
+      )
+    );
+    return {};
+  }
+};
+
+export const getIndexesMatchingIndexPatterns = async (
+  indices: string[],
+  services: AlertServices,
+  logger: Logger,
+  buildRuleMessage: BuildRuleMessage
+): Promise<Record<string, string[]>> => {
+  try {
+    const res = await indices.reduce(async (acc, indexPattern) => {
+      try {
+        const totalIndices: Array<Record<'index', string>> = await services.callCluster(
+          'cat.indices',
+          {
+            index: indexPattern,
+            format: 'json',
+          }
+        );
+        return { [indexPattern]: totalIndices.map((item) => item.index), ...(await acc) };
+      } catch (exc) {
+        logger.error(
+          buildRuleMessage(`[-] failed to cat indices: ${JSON.stringify(indices, null, 2)}`)
+        );
+        logger.error(buildRuleMessage(`Exception: ${exc}`));
+        return { ...(await acc) };
+      }
+    }, {} as Promise<Record<string, string[]>>);
+    return res;
+  } catch (exc) {
+    logger.error(
+      buildRuleMessage(`[-] failed to cat indices: ${JSON.stringify(indices, null, 2)}`)
+    );
+    logger.error(buildRuleMessage(`Exception: ${exc}`));
+    return Promise.resolve({});
+  }
+};
+
+export const findIndicesWithTimestampAndWithout = (
```

Comment:
This function combines the concrete index names from multiple index patterns into a single array of successes or failures, so we lose the association between a concrete index and the index pattern it comes from. Instead this could return an object where the keys are index patterns and the values contain arrays of the associated concrete indices that match or don't match.

We discussed on zoom how to refactor this function to maintain the association between index pattern and concrete index. The refactored design of this function, along with a refactored `checkIndexMappingsForTimestampFields` that returns the field mappings by index pattern rather than as a single group, will let us eliminate the complex regex logic that attempts to re-associate concrete indices to an index pattern in `timestampFieldCheck` [here](https://github.com/elastic/kibana/pull/83481/files#diff-333497786c31c4722a433be429554127c7eb40198062ef5004a057da5c0a23cfR906), [here](https://github.com/elastic/kibana/pull/83481/files#diff-333497786c31c4722a433be429554127c7eb40198062ef5004a057da5c0a23cfR925), and [here](https://github.com/elastic/kibana/pull/83481/files#diff-333497786c31c4722a433be429554127c7eb40198062ef5004a057da5c0a23cfR935)